pkg/network/usm/ebpf_gotls_helpers.go

```diff
@@ -8,15 +8,134 @@
 	"github.com/DataDog/datadog-agent/pkg/network/protocols/http/gotls"
+	libtelemetry "github.com/DataDog/datadog-agent/pkg/network/protocols/telemetry"
+	"github.com/DataDog/datadog-agent/pkg/network/usm/utils"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
 )
 
+// GoTLSBinaryInspector is a BinaryInspector that inspects Go binaries, dealing with the specifics of Go binaries
+// such as the argument passing convention and the lack of uprobes
+type GoTLSBinaryInspector struct {
+	structFieldsLookupFunctions map[bininspect.FieldIdentifier]bininspect.StructLookupFunction
+	paramLookupFunctions        map[string]bininspect.ParameterLookupFunction
+
+	// eBPF map holding the result of binary analysis, indexed by binaries'
+	// inodes.
+	offsetsDataMap *ebpf.Map
+
+	// binAnalysisMetric handles telemetry on the time spent doing binary
+	// analysis
+	binAnalysisMetric *libtelemetry.Counter
+
+	// binNoSymbolsMetric counts Golang binaries without symbols.
+	binNoSymbolsMetric *libtelemetry.Counter
+}
+
+// Ensure GoTLSBinaryInspector implements BinaryInspector
+var _ uprobes.BinaryInspector = &GoTLSBinaryInspector{}
+
+// Inspect extracts the metadata required to attach to a Go binary from the ELF file at the given path.
+func (p *GoTLSBinaryInspector) Inspect(fpath utils.FilePath, requests []uprobes.SymbolRequest) (map[string]bininspect.FunctionMetadata, bool, error) {
+	start := time.Now()
+
+	path := fpath.HostPath
+	f, err := os.Open(path)
+	if err != nil {
+		return nil, false, fmt.Errorf("could not open file %s, %w", path, err)
+	}
+	defer f.Close()
+
+	elfFile, err := elf.NewFile(f)
+	if err != nil {
+		return nil, false, fmt.Errorf("file %s could not be parsed as an ELF file: %w", path, err)
+	}
+
+	functionsConfig := make(map[string]bininspect.FunctionConfiguration, len(requests))
+	for _, req := range requests {
+		lookupFunc, ok := p.paramLookupFunctions[req.Name]
+		if !ok {
+			return nil, false, fmt.Errorf("no parameter lookup function found for function %s", req.Name)
+		}
+
+		functionsConfig[req.Name] = bininspect.FunctionConfiguration{
+			IncludeReturnLocations: req.IncludeReturnLocations,
+			ParamLookupFunction:    lookupFunc,
+		}
+	}
+
+	inspectionResult, err := bininspect.InspectNewProcessBinary(elfFile, functionsConfig, p.structFieldsLookupFunctions)
+	if err != nil {
+		if errors.Is(err, elf.ErrNoSymbols) {
+			p.binNoSymbolsMetric.Add(1)
+		}
+		return nil, false, fmt.Errorf("error extracting inspection data from %s: %w", path, err)
+	}
+
+	if err := p.addInspectionResultToMap(fpath.ID, inspectionResult); err != nil {
+		return nil, false, fmt.Errorf("failed adding inspection rules: %w", err)
+	}
+
+	elapsed := time.Since(start)
+	p.binAnalysisMetric.Add(elapsed.Milliseconds())
+
+	return inspectionResult.Functions, true, nil
+}
+
+// Cleanup removes the inspection result for the binary at the given path from the map.
+func (p *GoTLSBinaryInspector) Cleanup(fpath utils.FilePath) {
+	p.removeInspectionResultFromMap(fpath.ID)
```

Comment:
unlike the original implementation, seems like `removeInspectionResultFromMap` is being called and used only once (here)
in such a scenario, let's unify `removeInspectionResultFromMap` and `Cleanup` implementations